zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java

```diff
@@ -1849,13 +1924,23 @@ private void checkDataSizeLimit(byte[] data) {
   }
 
   public void watchForData(final String path) {
-    retryUntilConnected(new Callable<Object>() {
-      @Override
-      public Object call() throws Exception {
-        getConnection().exists(path, true);
-        return null;
+    watchForData(path, false);
+  }
+
+  private boolean watchForData(final String path, boolean skipWatchingNodeNotExist) {
+    try {
+      if (skipWatchingNodeNotExist) {
+        Stat stat = new Stat();
```

Comment:
https://zookeeper.apache.org/doc/r3.3.3/api/org/apache/zookeeper/ZooKeeper.html#getData(java.lang.String,%20boolean,%20org.apache.zookeeper.data.Stat)

They did not say this is guaranteed for you to pass null. What you state may be true for now, but not later. Or not earlier.

What is more, is there a way to check all the version of client having this behavior? 

Better to be defensive, seems to me. Saving such a constructor really means much. The trouble it can potentially bring to us is tremendous though.

Reply:
Zookeeper has unit tests everywhere for `getData()` with null passed in. For example: https://github.com/apache/zookeeper/blob/fe940cdd8fb23ba09684cefb73233d570f4a20fa/zookeeper-server/src/test/java/org/apache/zookeeper/test/AuthTest.java#L68

If our api could accept a null, we won't explicitly write it in Javadoc, right? I don't think such important behavior would be changed in zookeeper in future.

Reply:
Is there a way we can verify all the old existing zookeeper version can take null as out parameter? 

Relying unpublished behavior of API is not a good practices. We can also get our self into production issue in the future.